File: PYTHON/credit.py
from cs50 import get_string

card = get_string("Number: ")

# check length of number
length = len(card)
if length < 13 or length > 16:
    print("INVALID")

# list of digits
digits = []
for digit in card:
    digits.append(digit)

# get alternate digits
mults = []
singles = []
while length > 0:
    single_digit = digits[length - 1]
    singles.append(single_digit)
    length -= 1
    if length != 0:
        mult_digit = digits[length - 1]
        mults.append(mult_digit)
        length -= 1

# convert lists to numbers
mult_nums = [int(mult_num) for mult_num in mults]
single_nums = [int(single_num) for single_num in singles]

# get sum of single digits
single_sum = sum(single_nums)

# multiply digits
products = []
for mult_num in mult_nums:
    products.append(mult_num * 2)

# turn list of products into single number
prod = int(''.join(map(str, products)))

# add digits of products and get sum
prod_s = str(prod)
prod_l = len(prod_s)
prod_d = []
while prod_l > 0:
    dig = prod_s[prod_l - 1]
    prod_d.append(dig)
    prod_l -= 1
prod_digs = [int(dig) for dig in prod_d]
prod_sum = sum(prod_digs)

# get final sum and validate last digit
final_sum = single_sum + prod_sum
check = final_sum % 10

if check != 0:
    print("INVALID")
else:
    l = len(card)
    if l == 15 and digits[0] == "3" and digits[1] == "4" or digits[1] == "7":
        print("AMEX")
    elif l == 16 and digits[0] == "5" and digits[1] in ["1", "2", "3", "4", "5"]:
        print("MASTERCARD")
    elif l == 13 or l == 16 and digits[0] == "4":
        print("VISA")
    else:
        print("INVALID")

File: C/credit.c
#include <cs50.h>
#include <stdio.h>

long get_number(void);
long checksum(long number);

int main(void)
{
    // prompt user for input
    long number = get_number();

    // validate card
    int sum = checksum(number);
}

long get_number(void)
{
    long number;
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0);
    return number;
}

long checksum(long number)
{
    long last;
    int length = 0;
    int odd = 0;
    int even = 0;
    int even_digit = 0;
    int even_mult = 0;
    int odd_digit = 0;
    int odd_sum = 0;
    int even_sum = 0;
    int even_mult_digit, even_mult_split, even_mult_first;
    while (number > 0)
    {
        // find length of number
        length++;
        // find last number
        last = number % 10;
        // separate alternate numbers
        if (length % 2 == 1)
        {
            // add odd digits
            odd = odd * 10 + last;
            odd_digit = odd % 10;
            odd_sum = odd_digit + odd_sum;
        }
        else
        {
            // multiply odd digits and add products' digits
            even = even * 10 + last;
            even_digit = even % 10;
            even_mult_digit = even_digit * 2;
            even_mult_split = even_mult_digit % 10;
            even_mult_first = (even_mult_digit - even_mult_split) / 10;
            even_sum = even_mult_split + even_mult_first + even_sum;
        }
        number = number / 10;
    }
    // add all digits
    int sum = odd_sum + even_sum;
    // verify last digit of sum
    int sum_digit = sum % 10;
    if (sum_digit != 0)
    {
        printf("INVALID\n");
    }
    else if (length < 13 || length > 16)
    {
        printf("INVALID\n");
    }
    // determine company
    else if ((length == 13 || length == 16) && last == 4)
    {
        printf("VISA\n");
    }
    else if (length == 15 && last == 3 && (even_digit == 4 || even_digit == 7))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && last == 5 && (odd_digit == 1 || odd_digit == 2 || odd_digit == 3 || odd_digit == 4 || odd_digit == 5))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return sum;
}